### DISCUSSION TITLE
Staff balancing

### DISCUSSION THREAD

**Discussion by opekope2** [opekope2/StaffMod#6] [General]
This is the discussion for balancing general properties of the staff, like attack damage and speed (equip time) and ways of obtaining it.

**Comment by opekope2** [comment]
Blue has +1 reach while holding the staff at [1:43](https://youtu.be/vlVUVsx6BYY?t=103).

**Reply by opekope2** [reply]
In **The Raid**, an armorer, a cleric, and a fletcher create the lapis staff. However, I see the weaponsmith more fitting. I'll think about the toolsmith, feel free to convince me. I'll wait with the lapis staff until Mojang merges villager trade rebalance into the main game, so it's not just an experiment.
The trade will of course be a master level trade with a plains villager (in **The Raid**, the village looks like a meadow village, which is just a plains village).
The problem with the cleric trade is that a villager would work for free (no emeralds involved, because the staff and the lapis would fill up the trade slots). In the vanilla game, a villager always trades with emeralds.

**Reply by opekope2** [reply]
The way I'll make King Orange's staff obtainable is by crafting two staff rods, and its top part together, then in a smithing table, adding redstone to it. I have the same crafting mechanic idea for the villager-forged staff, which means the weaponsmith would only trade the top part of the staff, and the player would craft the staff, and add lapis using the smithing table. Yes, this makes the mod not fully canonically accurate, but I have to nerf other things to keep the gameplay balanced.

**Reply by BoB10011** [reply]
Hey I think the Mushroom staff could be like a shield like from the episode

**Reply by BoB10011** [reply]
also for the *any* ice staff you can make it so that it places ice on mobs and in general where the player looks and gives mobs the 100% slowness effect

**Reply by BoB10011** [reply]
for the piston staff, you could make a simple piston pushing animation for it a it has a lot of knockback

**Reply by charmander9695** [reply]
I see a problem and it has to do with the block variety like wood. It controls all in that section (oak wood, planks, stairs and etc.). those type's of blocks are probable going to be the hardness to code

**Reply by Devbuilder10011** [reply]
Also, why is there durability? The staff never broke or was breaking the only staff that broke was the lapis one

**Reply by opekope2** [reply]
Every tool and weapon has a durability in the vanilla game, so I added a high durability (~2.5x netherite pickaxe), for survival balancing

**Comment by opekope2** [comment]
Change staff stats in 0.2.0-beta

* Increase damage 1->4 HP
* Decrease attack speed 4->2 (still faster than any vanilla weapons)
* Increase DPS 4->8 HP

**Comment by charmander9695** [comment]
i do have a question about Advanced block functionalities inside staff item in phase 2

**Reply by opekope2** [reply]
"Advanced block functionality": basically client side profile, which is uploaded to the server when an item is inserted into the staff, and saved as an NBT. Profiles can be changed with a command (and a macro mod). For example, one may have an attack and defense brewing stand profile, throwing specific potions. Crafting table with preferred recipes (if I choose this implementation) is similar.

**Reply by opekope2** [reply]
@charmander9695 [just ask it](https://dontasktoask.com), if I didn't answer it with my previous comment

**Reply by charmander9695** [reply]
ok thank you for the info

**Comment by charmander9695** [comment]
a way of obtaining the block staff in survival can be you making it like in ep 30 ( 7:56 to 9:07 ) but another way is to trade for it like in ep 28
( 1:28 to 3:18 ) or you can do is make some parts traded and make some parts makeable

**Reply by opekope2** [reply]
I'll eventually add Yellow's old staff, which would be obtainable with a villager trade, and then the player would need to add lapis to it to make it functional (not with an enchanting table, I think it would be a cool use for the smithing table)

**Reply by opekope2** [reply]
The current staff would be the craftable one with expensive resources (because it's powerful). I'm not sure if I want to add the non-functional (failed) staffs

**Comment by opekope2** [comment]
I also need some Minecraft names for the different staffs. I wouldn't use names "Yellow's staff", "Yellow's old staff", or "King Orange's staff", because there are multiple of these, and stick figures don't exist in Minecraft (without other mods).
Some names I can think of are "Perfect staff", "Enchanted staff" ("Unenchanted staff" sounds meh). I'd like to hear your creative ideas

**Reply by charmander9695** [reply]
why not just call kings block staff the power staff

**Comment by charmander9695** [comment]
speaking of obtaining the staff I have an image that could help 
trident is the block holder and Redstone block is half the stick

![2024-01-21_20 33 12](https://github.com/opekope2/StaffMod/assets/156858642/92c81df1-f8e2-453d-a01e-efcb5fbb1bfa)
hope you see it

**Reply by opekope2** [reply]
I'll consider your idea when implementing crafting

**Reply by charmander9695** [reply]
ok thank you for considering the thought

**Reply by opekope2** [reply]
The blaze rod gave me an idea. It's a rod, so it makes sense if it's in the crafting recipe. End rod is also a rod. The third item should be one from the overworld, so the player has to travel to all 3 dimensions. Netherite is a good idea, to make the staff expensive (because it's versatile, and give another use to the netherite ingot), but all components can be obtained in the nether. I think netherite ingots only in the "staff head" can work. A stick kinda looks like a rod, but it's too common, and can be obtained in the nether. The one I found to be challenging to obtain, yet it has not many uses, is the echo shard.
The final crafting recipe (2 rods + "head") is an obvious recipe, so I'll definitely implement it.
[And there would be an additional step, where I can include redstone or another material](https://github.com/opekope2/StaffMod/discussions/6#discussioncomment-8354138)

**Reply by opekope2** [reply]
Mojang just added the breeze rod in the latest snapshot, which is basically the 3rd rod I was looking for

**Comment by opekope2** [comment]
I'm thinking about a multi-step crafting for the staff:
1. Craft the "staff rods" and the "staff head" (not the final names)
2. Craft the staff
3. In a smithing table, infuse a material (like redstone or lapis) to the staff to make it functional

The material would be defined by the "staff head" (King Orange's "Perfected" Staff would be redstone, Yellow's "NPC-crafted" Staff would be lapis lazuli, based on the staff's color)

I *may* add multiple material options to "staff rods" than [echo shard, as described here](https://github.com/opekope2/StaffMod/discussions/6#discussioncomment-8354081), which would affect the staff's power (more common materials would debuff it, like less durability, more cooldown, slower shooting, digging less blocks, etc)

I have some ideas about the generation of the smithing template:
* An end ship: it's not that common, and end rods can be crafted (no end city raiding is required)
* Villager trade (would work with the rebalance if/when implemented by Mojang)
* Wandering trader (maybe too hard)
* Archeology

**Comment by opekope2** [comment]
How much durability should the staff have?

A breaking mechanic, which came to my mind, is that the staff would break into parts when broken, so the user would get back the item in the, the "staff head", and the two "staff rods", but the enchantments would be lost, and the staff would need to be crafted again. I don't remember seeing a staff breaking in the animations, so this is just an idea, which may never be implemented.

Or just use mending and pay attention to the durability, because some items would drain it faster (like the upcoming netherite block)

**Comment by lapisenjoyer** [comment]
- i think the staff should have an energy meter that is used up with use, but instead of breaking, it needs to be re-infused. i have a few ideas for some staves too: 
- redstone block: left click activates redstone components as if there were a button there. holding right click fires a powerful laser, this would probably drain energy really fast.
- lapis block: hold right click to gain a limited flight ability. the max height you can fly above the ground would be linked to exp levels. i personally think it should be one block for every three levels.
- anvil change: adding an anvil to a staff midair drops you down quicker and does an aoe slam. additionally, the anvils slow still partially effects the player when in the inventory.
- noteblock: left click to trap mobs directly in front of the user on a notation chart. right click to fire music notes that deal damage. damage is normally small but is increased against mobs on the notation chart.

**Reply by opekope2** [reply]
The staff would break into its "head" and two "rods", and would drop its item, which can be crafted again to a staff, then the material needs to be infused again

**Reply by opekope2** [reply]
Note block: you basically read my mind. At least partially. Another feature I plan is to "just shoot notes"

**Reply by opekope2** [reply]
Anvil: 
* slowing down while in the inventory would need some changes to the vanilla codebase, and I try to only minimally touch vanilla mechanics
* Holding an anvil staff to fall down quicker can be done in Minecraft 1.20.5+, but that's not physically accurate
* AOE slam is an interesting idea. Instead of adding an anvil midair, I would hold down RMB, and the AOE damage would be calculated based on the fall damage

**Reply by opekope2** [reply]
Lapis is a mutifunctional block. [TSC can fly, and change blocks to have enchantment glint](https://youtu.be/vlVUVsx6BYY?t=142). [Blue can change flying blocks and control them](https://youtu.be/vlVUVsx6BYY?t=964). I'm not sure how I would implement the "changing and controlling the affected blocks" functionality.

**Reply by opekope2** [reply]
Redstone block: 
* my idea is that the laser activates the components while it points at said component (continuously power redstone dust, continuously pulse observer). Left clicks could be used to 1-tick redstone dust for example, I'll think about this.
* block slicing is not something I want to implement
* the laser would blow up ignited TNT, [as seen in the bee fight scene](https://youtu.be/vlVUVsx6BYY?t=743), and damage entities

**Reply by opekope2** [reply]
The durability bar as an energy bar suffices IMO. Some blocks would obviously drain it faster than others, and unbreaking could be used to decrease this amortization. Mending would be a must, or use anvils to repair it.

**Reply by lapisenjoyer** [reply]
i completely understand if you want to keep a standard durability system. sorry if i overstepped. i only have one more balancing concern then. in lucky block staff, we see the stick's swapping staves constantly, despite having fairly powerful ones from the start. i'm not the developer you are, so this is up to you, but i think the staffs should encourage a play-style of rapid swapping. in the current system (and likely even after some of the mechanics you discussed get added) it is really easy to use a magma or tnt staff and become untouchable

**Reply by charmander9695** [reply]
the Lapis block did half the work. blue also used the brewing stand to change them to chests

**Reply by opekope2** [reply]
I didn't catch that, thanks for telling. Brewing stand changed some monsters to blocks before, so it makes sense

**Reply by opekope2** [reply]
Mojang just added the mace in the latest snapshot, which gives is almost an AOE slam weapon. Now I'm more positive about adding it to the anvil staff

**Reply by opekope2** [reply]
Anvil staff AOE demo:

https://github.com/opekope2/StaffMod/assets/20487556/e198e11e-6b1b-4933-befa-0ac38419b69d

Coming in 0.15.0

**Reply by charmander9695** [reply]
a thought is just in my mind that maybe is broken but the staff allows anything to happen without anything bad to happen (except the lucky block) so putting limits on is may be hard to do.

**Comment by opekope2** [comment]
I nerfed snow block, magma block, and TNT staff in 0.13.1-beta. These will only shoot their respective projectile/explosive, if the attack indicator is fully charged. Right click rapid fire still works using snow and magma block if the attack indicator is fully charged.
Let me know your opinions about this change.

**Comment by TeboMC** [comment]
The Dispenser Staff can have a enchantment where every level the staff has and it has a dispenser, it adds 1 dispenser slot. the enchant could be called "Dispenser Slot" and goes up to level 9.

**Comment by TeboMC** [comment]
Emerald Block Staff Has Too High Damage. It Nearly is as strong as the Netherite Block Staff. You should update the mod to 1.21.4 and add dispenser staff btw.

**Reply by opekope2** [reply]
The Netherite Block Staff has 2.5x damage (10❤️) while the Emerald Block Staff has (4❤️). The default damage is (2❤️).
I don't disagree that 2-hitting (with crits) is strong, but we're talking about a massive block of emerald.
